tests: subsys: secure_storage: Erase flash before write once test

petejohanson-adi opened the pull request:
Where possible (i.e. not running in NS), erase the flash before testing the write once secure storage API.

This is based on feedback here: https://github.com/zephyrproject-rtos/zephyr/pull/97529#pullrequestreview-3334815184

Note: This does *not* help with NS targets that can't directly access the flash and erase it. @tomi-font I'd really appreciate if you could offer some ideas for how to address that problem other than the one proposed in #97529

petejohanson-adi committed:
tests: subsys: secure_storage: Erase flash before write once test

Where possible (i.e. not running in NS), erase the flash before testing the
write once secure storage API.

Signed-off-by: Pete Johanson <pete.johanson@analog.com>

dsseng commented:
Are NS targets even supposed to use secure_storage? I believe if you cannot access the flash from NS, then you would typically use PSA ITS/PS provided by the secure world, thus not needing Zephyr's implementation

dsseng commented on tests/subsys/secure_storage/psa/its/src/main.c:
```diff
@@ -4,7 +4,49 @@
+#if defined(CONFIG_FLASH_HAS_NO_EXPLICIT_ERASE)
+	const struct flash_parameters *fparam = flash_get_parameters(fdev);
+
+	if (!(flash_params_get_erase_cap(fparam) & FLASH_ERASE_C_EXPLICIT)) {
```
Maybe in case we have no erase operation (e.g. nRF54 RRAM and MRAM) we should fully overwrite the region with zeroes or ones? That would also erase the data to make sure we start from a known-empty state.

native_sim can also simulate such NVM model IIRC, so it should be testable even without such HW.

dsseng commented on tests/subsys/secure_storage/psa/its/src/main.c:
```diff
@@ -4,7 +4,49 @@
+
+	for (int i = 0; i < MAX_NUM_PAGES && address < FLASH_ERASE_END_ADDR; i++) {
+		rc = flash_get_page_info_by_offs(fdev, address, &page_info);
+
```
non-blocking nit to unify the code style here

```suggestion
```

petejohanson-adi commented:
In our particular case, we have a possible convoluted path of secure_storage -> settings -> settings ITS backend.

dsseng commented:
> In our particular case, we have a possible convoluted path of secure_storage -> settings -> settings ITS backend.

Yes, nothing prevents this from working, but perhaps only running this test without TF-M is sufficient for the coverage purposes

tomi-font commented on tests/subsys/secure_storage/psa/its/src/main.c:
```diff
@@ -4,7 +4,49 @@
+
+#if !defined(CONFIG_BUILD_WITH_TFM) && defined(CONFIG_FLASH_HAS_EXPLICIT_ERASE)
+
+#define MAX_NUM_PAGES 2
```
How did you come up with this value? Is it even needed?

tomi-font reviewed (commented):
@dsseng is right in that NS targets are not meant to use Secure Storage:
https://github.com/zephyrproject-rtos/zephyr/blob/f1c2ce4f4026bd5b212c6e4cc7d754aba69b189e/subsys/secure_storage/Kconfig#L4-L6 However this test has a scenario for TF-M in order to test API compatibility between TF-M and Secure Storage (as all the test does is use the PSA ITS API):
https://github.com/zephyrproject-rtos/zephyr/blob/f1c2ce4f4026bd5b212c6e4cc7d754aba69b189e/tests/subsys/secure_storage/psa/its/testcase.yaml#L67-L71 So it makes sense to also consider TF-M here.


> In our particular case, we have a possible convoluted path of secure_storage -> settings -> settings ITS backend.

This is really weird... Can you tell me more about your use case? Why wouldn't you just directly be using the PSA ITS API exposed by TF-M, and how can you even manage to do this as there must be symbol clashes?

Other than that, thanks for the PR, I tested and it works on the nRF9151 and on `native_sim` with the Settings backend. Although not:
- On the nRF54L15, probably because of https://github.com/zephyrproject-rtos/zephyr/pull/99701/files#r2543678217; could you also make it work there?
- With the ZMS backend. In that case (on `native_sim`) the return value of `psa_its_get_info()` is not `PSA_SUCCESS`, which makes the test fail. This is maybe because the erase operation messes up ZMS or that the mounting needs to be done _after_ erasing the flash: https://github.com/zephyrproject-rtos/zephyr/blob/bd9dc902e180c967dfb914a3de3ae6628b76ef43/subsys/secure_storage/src/its/store/zms.c#L22-L34 Maybe having the flash erase as a `SYS_INIT()` hook that would run before that of ZMS would work.

> Note: This does not help with NS targets that can't directly access the flash and erase it. @tomi-font I'd really appreciate if you could offer some ideas for how to address that problem other than the one proposed in https://github.com/zephyrproject-rtos/zephyr/pull/97529

I'm afraid there's no way around that, for NS targets you will just need to run twister/west with the erase option. It would be a pretty big security hole if the NS application could just erase parts of the secure flash. I mean I imagine it must be possible some way to achieve that but it must be all but trivial.

dsseng commented:
> I'm afraid there's no way around that, for NS targets you will just need to run twister/west with the erase option. It would be a pretty big security hole if the NS application could just erase parts of the secure flash. I mean I imagine it must be possible some way to achieve that but it must be all but trivial.

Can be potentially hacked around by making TF-M skip setting security attributes to the flash controller (needs a test-specific patch). It's still not really worth it for the test though it seems.

And on properly secured hardware that's the goal of TF-M to prevent NSPE from accessing Secure hardware, so there's no universal solution without hacking on TF-M.

@petejohanson-adi I still fail to grasp what do you test in such a case? If you use Secure Storage, then it writes to somewhere accessible from NSPE, right? Then that region should also be erasable/rewriteable by NSPE.

If you meant Secure Storage on top of ITS provided by TF-M (yes, I now realized it shouldn't even compile due to TF-M runtime library providing the same symbol names as secure_storage), please take a look here, maybe you should wipe ITS entries, and not the backing Flash: https://github.com/zephyrproject-rtos/zephyr/pull/94881/files#diff-93455f3e147570503f332b1446633676ffb36a58f958179f334273ca6b30c78e

petejohanson-adi commented:
> @dsseng is right in that NS targets are not meant to use Secure Storage:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/f1c2ce4f4026bd5b212c6e4cc7d754aba69b189e/subsys/secure_storage/Kconfig#L4-L6
> 
> However this test has a scenario for TF-M in order to test API compatibility between TF-M and Secure Storage (as all the test does is use the PSA ITS API):
> https://github.com/zephyrproject-rtos/zephyr/blob/f1c2ce4f4026bd5b212c6e4cc7d754aba69b189e/tests/subsys/secure_storage/psa/its/testcase.yaml#L67-L71
> 
> So it makes sense to also consider TF-M here.
> > In our particular case, we have a possible convoluted path of secure_storage -> settings -> settings ITS backend.
> 
> This is really weird... Can you tell me more about your use case? Why wouldn't you just directly be using the PSA ITS API exposed by TF-M, and how can you even manage to do this as there must be symbol clashes?

My apologies... That message was totally in error, from me incorrect quick assessment of the test case that was running, and getting mixed up with the failures we've been chasing on the non-NS targets. It is in fact the scenario you're highlighting, with the TFM test to verify compatibility.

> 
> Other than that, thanks for the PR, I tested and it works on the nRF9151 and on `native_sim` with the Settings backend. Although not:
> 
> * On the nRF54L15, probably because of https://github.com/zephyrproject-rtos/zephyr/pull/99701/files#r2543678217; could you also make it work there?

I'll take a look, thanks.

> * With the ZMS backend. In that case (on `native_sim`) the return value of `psa_its_get_info()` is not `PSA_SUCCESS`, which makes the test fail. This is maybe because the erase operation messes up ZMS or that the mounting needs to be done _after_ erasing the flash: https://github.com/zephyrproject-rtos/zephyr/blob/bd9dc902e180c967dfb914a3de3ae6628b76ef43/subsys/secure_storage/src/its/store/zms.c#L22-L34
>    Maybe having the flash erase as a `SYS_INIT()` hook that would run before that of ZMS would work.

Yeah, I was a bit worried about that possible case. I'm fine moving it to an earlier `SYS_INIT` hook, that's probably more reliable across the board.

> 
> > Note: This does not help with NS targets that can't directly access the flash and erase it. @tomi-font I'd really appreciate if you could offer some ideas for how to address that problem other than the one proposed in #97529
> 
> I'm afraid there's no way around that, for NS targets you will just need to run twister/west with the erase option. It would be a pretty big security hole if the NS application could just erase parts of the secure flash. I mean I imagine it must be possible some way to achieve that but it must be all but trivial.

Thanks, that was exactly my take as well, but unfortunately, running a whole twister suite with the `--erase` option for *every* test has a non-trivial impact to the total test time (the bulk erase time adds up), which is why in #97529 we had tried to suggest a full erase *just for this test*. I understand why many that may not be desirable for *all* targets... Thoughts adding that erase to the runners for this test, but conditional on a CMake define that folks can use to opt into that behavior? @tomi-font ?

tomi-font commented:
> Thanks, that was exactly my take as well, but unfortunately, running a whole twister suite with the --erase option for every test has a non-trivial impact to the total test time (the bulk erase time adds up), which is why in https://github.com/zephyrproject-rtos/zephyr/pull/97529 we had tried to suggest a full erase just for this test. I understand why many that may not be desirable for all targets... Thoughts adding that erase to the runners for this test, but conditional on a CMake define that folks can use to opt into that behavior? @tomi-font ?

I see, I wouldn't be against for example adding a Kconfig option to that test to allow opting in if the approach proposed in https://github.com/zephyrproject-rtos/zephyr/pull/97529 would be acceptable in that case. @nordicjm @de-nordic thoughts?

tomi-font commented on tests/subsys/secure_storage/psa/its/src/main.c:
```diff
@@ -4,7 +4,62 @@
+
+#if defined(CONFIG_FLASH_HAS_NO_EXPLICIT_ERASE)
+		/* Lacking an explicit erase, just write all ones to the pages */
+		uint8_t empty[page_info.size];
```
VLAs are not allowed

petejohanson-adi replied:
Ok, I'll adjust and use a fixed size array and iterate.

petejohanson-adi replied:
Adjusted.

tomi-font commented:
> I am ok with adding Kconfig to control this, I am not ok with directly using Flash API kconfig in code, it can be used in test Kconfig.

Do you mean to do
```
config MY_TEST_KCONFIG
        def_bool FLASH_HAS_NO_EXPLICIT_ERASE
```
? If so, why?

de-nordic commented:
> > I am ok with adding Kconfig to control this, I am not ok with directly using Flash API kconfig in code, it can be used in test Kconfig.
> 
> Do you mean to do
> 
> ```
> config MY_TEST_KCONFIG
>         def_bool FLASH_HAS_NO_EXPLICIT_ERASE
> ```
> 
> ? If so, why?

Because the Flash API Kconfig belongs to Flash API. If I, or anyone else, needs to change that Kconfig for some reason I would have to go through a lot of subsystems code to fix it being used there to. If the Kconfig is only driving other Kconfigs then I only need to go through Kconfigs.
Additionally, this Kconfig is just informing that there is certain type of a device in system; subsystem may or may not support it, or have specific option to select support for such a device, and that options should be dependent on the Flash API Kconfig; because that option actually controls behavior and paths within subsystem, no Flash API.

petejohanson-adi commented:
> > Do you mean the flash subsystem may not support it or some other subsystem that makes use of the flash API? In the case of this PR it's purely about a test that is directly using the flash API to manipulate the NVM.
> 
> Example from Stream Flash
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/c1d16c0555d667c9a3650c0ff69aba18f858b8cb/subsys/storage/stream/Kconfig#L32-L37
> 
> the above option only appears if there is at least one device with explicit erase. User may select it or not, depending whether user intends to use the subsystem on device with this characteristic.
> 
> If you assume that Kconfig is itself a separate configuration module, consisting of sub-modules steering subsystems that will be compiled in, then these sub-modules should only control sub-modules they have been developed for. So even if flash Kconfigs may have direct impact on other subsystem configuration, for example Stream Flash, the should not have direct usage in subsystems they have not been designed to control.
> 
> This is my opinion and I may be wrong here, but imho it looks better in code if it is controlled, even in compilation, by identifiers directly intended for that subsystem/module.

I've locally adjusted the code to just check `flash_params_get_erase_cap(fparam)` for the `FLASH_ERASE_C_EXPLICIT` flag, and then deciding based on that how in the test to try to clear things out. Is this more aligned with your expectations? It's no longer using those "internal" Kconfig symbols to determine how to proceed.

petejohanson-adi commented:
Ok, I couldn't get this happy on the nrf52l15 here... For now, I've adjusted the "no explicit erase" to use flash writes only if the new tests-specific Kconfig is enabled. I'm open to other ideas here.

@de-nordic adjusted to the runtime check of the caps in this version as well, if you can revisit.

Thanks.

tomi-font commented on tests/subsys/secure_storage/psa/its/Kconfig:
```diff
@@ -0,0 +1,10 @@
+# Private config options for its test
+
+# Copyright (c) 2025 Analog Devices, Inc
+# SPDX-License-Identifier: Apache-2.0
```
Copyright first

tomi-font commented on tests/subsys/secure_storage/psa/its/src/main.c:
```diff
@@ -4,7 +4,79 @@
+/* Low priority to ensure we run after any flash drivers are initialized */
+SYS_INIT(erase_flash, POST_KERNEL, 100);
+
+#endif /* !defined(CONFIG_BUILD_WITH_TFM) && defined(CONFIG_FLASH_PAGE_LAYOUT) */
```
```suggestion
#endif /* !CONFIG_BUILD_WITH_TFM && CONFIG_FLASH_PAGE_LAYOUT */
```

tomi-font commented on tests/subsys/secure_storage/psa/its/src/main.c:
```diff
@@ -4,7 +4,79 @@
+					return rc;
+				}
+
+				to_clear -= FLASH_MEMSET_CLEAR_SIZE;
```
Is `page_info.size` guaranteed to always be a multiple of `FLASH_MEMSET_CLEAR_SIZE`? I would change `to_clear` to be a `ssize_t` to be on the safe side.

tomi-font reviewed (changes requested):
I tested on `native_sim` and on a re-run of `secure_storage.psa.its.secure_storage.store.zms` we get:
```
START - test_write_once_flag

    Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/secure_storage/psa/its/src/main.c:194: secure_storage_psa_its_test_write_once_flag: (ret not equal to PSA_SUCCESS)
Has the flash been erased since this test ran? -133
 FAIL - test_write_once_flag in 0.000 seconds
```
An interesting and maybe relevant thing is that on the first run there is
```
Setting 1000 at FC000 to 0xFF
Setting 1000 at FD000 to 0xFF
Setting 1000 at FE000 to 0xFF
Setting 1000 at FF000 to 0xFF
```
while on the second run (which fails) there is only
```
Setting 1000 at FC000 to 0xFF
```

tomi-font commented on tests/subsys/secure_storage/psa/its/src/main.c:
```diff
@@ -4,7 +4,79 @@
+
+			memset(clear_vals, fparam->erase_value, FLASH_MEMSET_CLEAR_SIZE);
+
+			TC_PRINT("Setting %zX at %lX to 0x%02X\n", page_info.size,
```
be consistent on the prints, hex or decimal but stick to the same

petejohanson-adi commented:
> I tested on `native_sim` and on a re-run of `secure_storage.psa.its.secure_storage.store.zms` we get:
> 
> ```
> START - test_write_once_flag
> 
>     Assertion failed at WEST_TOPDIR/zephyr/tests/subsys/secure_storage/psa/its/src/main.c:194: secure_storage_psa_its_test_write_once_flag: (ret not equal to PSA_SUCCESS)
> Has the flash been erased since this test ran? -133
>  FAIL - test_write_once_flag in 0.000 seconds
> ```
> 
> An interesting and maybe relevant thing is that on the first run there is
> 
> ```
> Setting 1000 at FC000 to 0xFF
> Setting 1000 at FD000 to 0xFF
> Setting 1000 at FE000 to 0xFF
> Setting 1000 at FF000 to 0xFF
> ```
> 
> while on the second run (which fails) there is only
> 
> ```
> Setting 1000 at FC000 to 0xFF
> ```

Ok, reproduced that here... Was a regression I missed when I moved away from the VLA, AFAICT.

This should now be working, and several of your review items are now moot, as I've discovered and moved to `flash_flatten`, which seems to do exactly what we want for each different case, and allows us not to worry about the specifics. Tested on `native_sim/native/64` (with and without `--extra-args CONFIG_FLASH_SIMULATOR_EXPLICIT_ERASE=n`) and several max32 targets, and it works as expected. nrf54l15 still has some funkiness, so I've not attempted to included it in this fix.

tomi-font reviewed (commented):
Nice update. I tested on 54L15 as well and now with `CONFIG_TESTS_SECURE_STORAGE_PSA_ITS_ERASE_WITH_FLASH_WRITE=y` it actually works with both the ZMS and Settings backends.
It seems that we could try to always erase, on all platforms. So ideally we could get rid of the new Kconfig option and just always call `flash_flatten()`.

de-nordic commented:
> Nice update. I tested on 54L15 as well and now with `CONFIG_TESTS_SECURE_STORAGE_PSA_ITS_ERASE_WITH_FLASH_WRITE=y` it actually works with both the ZMS and Settings backends. It seems that we could try to always erase, on all platforms. So ideally we could get rid of the new Kconfig option and just always call `flash_flatten()`.

Yes. If your goal evolved to remove data anyway, regardless of device, then simply calling `flash_flatten` will have all the logic you need to call proper function to do so.

petejohanson-adi commented on tests/subsys/secure_storage/psa/its/src/main.c:
```diff
@@ -4,7 +4,43 @@
 
+#include <zephyr/types.h>
+#include <zephyr/storage/flash_map.h>
+#include <zephyr/drivers/flash.h>
```
Adjusted. Better?

tomi-font reviewed (dismissed):
Very nice!

dsseng commented on tests/subsys/secure_storage/psa/its/src/main.c:
```diff
@@ -1,10 +1,36 @@
+static int erase_flash(void)
+{
+	int rc;
+	const struct device *const fdev = DEVICE_DT_GET(DT_CHOSEN(zephyr_flash_controller));
```
a visual nit

```suggestion
/* The flash must be erased after this test suite is run for the write-once entry test to pass. */
#if !defined(CONFIG_BUILD_WITH_TFM) && defined(CONFIG_FLASH_PAGE_LAYOUT)
static int erase_flash(void)
{
	const struct device *const fdev = DEVICE_DT_GET(DT_CHOSEN(zephyr_flash_controller));
	int rc;
```

dsseng reviewed (approved):
Thanks, looks great now!
